.github/workflows/build.yml

```
@@ -0,0 +1,305 @@
+        include:
+          - namespace: test
+            network: testnet
+          - namespace: prod
+            network: mainnet
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          submodules: recursive
+
+      # CACHE_VERSION secret is 'date --iso-8601=minutes' and is used to invalidate cache if needed
+      - name: Cache Build Binaries
+        id: artifact_cache
+        uses: actions/cache@v3
+        with:
+          path: |
+            build_artifacts
+          key: ${{ runner.os }}-${{ matrix.network }}-${{ secrets.CACHE_VERSION }}-build-cargo-artifacts-${{ hashFiles('**/*.rs', '**/*.proto', '**/Cargo.toml')}}
+
+      - name: Cache Cargo
+        if: steps.artifact_cache.outputs.cache-hit != 'true'
+        id: cargo_cache
+        uses: actions/cache@v3
+        with:
+          path: |
+            /opt/cargo/bin/
+            /opt/cargo/registry/index/
+            /opt/cargo/registry/cache/
+            /opt/cargo/git/db/
+            target/
+          key: ${{ runner.os }}-${{ secrets.CACHE_VERSION }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+
+      - name: Consensus SigStruct
+        if: steps.artifact_cache.outputs.cache-hit != 'true'
+        run: |
+          CONSENSUS_SIGSTRUCT_URI=$(curl -s https://enclave-distribution.${{ matrix.namespace }}.mobilecoin.com/production.json | grep consensus-enclave.css | awk '{print $2}' | tr -d \" | tr -d ,)
+          (cd /var/tmp && curl -O https://enclave-distribution.${{ matrix.namespace }}.mobilecoin.com/${CONSENSUS_SIGSTRUCT_URI})
+
+      - name: Ingest SigStruct
+        if: steps.artifact_cache.outputs.cache-hit != 'true'
+        run: |
+          INGEST_SIGSTRUCT_URI=$(curl -s https://enclave-distribution.${{ matrix.namespace }}.mobilecoin.com/production.json | grep ingest-enclave.css | awk '{print $2}' | tr -d \" | tr -d ,)
+          (cd /var/tmp && curl -O https://enclave-distribution.${{ matrix.namespace }}.mobilecoin.com/${INGEST_SIGSTRUCT_URI})
+
+      - name: Cargo Build
+        if: steps.artifact_cache.outputs.cache-hit != 'true'
+        run: |
+          cargo build --release
+
+      - name: Copy binaries to cache folder
+        if: steps.artifact_cache.outputs.cache-hit != 'true'
+        run: |
+          mkdir -pv build_artifacts/${{ matrix.network }}/bin
+          cp /var/tmp/*.css build_artifacts/${{ matrix.network }}
+          cp target/release/full-service build_artifacts/${{ matrix.network }}/bin/
+
+      # Create and Upload an Artifact on Push and Not a Tag
+      - name: Create Artifact
+        run: |
+          mkdir -pv artifact
+          cd artifact && tar -czvf ${{ github.sha }}-linux-${{ matrix.network }}.tar.gz -C ../build_artifacts/${{ matrix.network }}/ .
+
+      - name: Upload Artifact
+        uses: actions/upload-artifact@v3
+        with:
+          name: full-service_linux_${{ matrix.network }}
+          path: artifact/${{ github.sha }}-linux-${{ matrix.network }}.tar.gz
```

Optional/Equivalent: Put the SHA at the end
```suggestion
          cd artifact && tar -czvf linux-${{ matrix.network }}-${{ github.sha }}.tar.gz -C ../build_artifacts/${{ matrix.network }}/ .

      - name: Upload Artifact
        uses: actions/upload-artifact@v3
        with:
          name: full-service_linux_${{ matrix.network }}
          path: artifact/linux-${{ matrix.network }}-${{ github.sha }}.tar.gz
```